src/main/scala/Checkout.scala

```diff
@@ -1,129 +1,161 @@
 
-  // Char SKUs mapping to offers
-  val offers: Map[Char, (Int, Int)] = Map[Char, (Int, Int)](
-    // SKU -> (x for y)
-    'A' -> (3, 130),
-    'B' -> (2, 45)
-  )
+  val itemA = Item("A", 50, Option(Offer(3, 130)))
+  val itemB = Item("B", 30, Option(Offer(2, 45)))
+  val itemC = Item("C", 20, None)
+  val itemD = Item("D", 15, None)
 
-  val exampleShop = List('A', 'B', 'C', 'D', 'A', 'A', 'B', 'D', 'D', 'D')
+  val testShop: List[Item] =
+    List(itemA, itemB, itemC, itemD, itemA, itemA, itemB, itemD, itemD, itemD)
 
 }
 
 object Checkout extends App {
 
   import Inventory._
 
+  print(Console.UNDERLINED)
   println("Checkout Kata\n")
+  print(Console.RESET)
 
   /**
     * Calculates subtotal with a recursive definition
     *
     * @param items Products to be purchased
-    * @param stock Char SKUs mapping to prices
-    * @return Item subtotal
+    * @return item subtotal
     */
-  def calculateSubtotalWithRecursion(items: List[Char], stock: Map[Char, Int]): Int =
+  def calculateSubtotalWithRecursion(items: List[Item]): Int =
     if (items.isEmpty)
       0
     else {
-      val item: Char = items.head
-      val price: Int = stock(item)
-      val rest: List[Char] = items.tail
-      price + calculateSubtotalWithRecursion(rest, stock)
+      val price: Int = items.head.price
+      val rest: List[Item] = items.tail
+      price + calculateSubtotalWithRecursion(rest)
     }
 
   /**
     * Calculates subtotal with pattern matching
     *
     * @param items Products to be purchased
-    * @param stock Char SKUs mapping to prices
     * @return Item subtotal
     */
-  def calculateSubtotalWithPM(items: List[Char], stock: Map[Char, Int]): Int =
+  def calculateSubtotalWithPM(items: List[Item]): Int =
     items match {
       case Nil =>
         0
       case h :: t =>
-        stock(h) + calculateSubtotalWithPM(t, stock)
+        h.price + calculateSubtotalWithPM(t)
     }
 
+  /**
+    * Calculates subtotal using tail recursion (wrapper function)
+    *
+    * @param items Products to be purchased
+    * @return Item subtotal
+    */
+  def calculateSubtotalWithTailRecursion(items: List[Item]): Int = {
+
+    /**
+      * Calculates subtotal using tail recursion
+      *
+      * @param items Products to be purchased
+      * @param acc   Accumulator
+      * @return Keeps track of the accumulator
+      */
+    @tailrec
+    def tailRecursion(items: List[Item], acc: Int): Int =
+      items match {
+        case Nil =>
+          acc
+        case h :: t =>
+          tailRecursion(t, acc + h.price)
+      }
+
+    tailRecursion(items, 0)
+  }
+
   /**
     * Calculates subtotal with a fold
     *
     * @param items Products to be purchased
-    * @param stock Char SKUs mapping to prices
     * @return Item subtotal
     */
-  def calculateSubtotalWithFold(items: List[Char], stock: Map[Char, Int]): Int =
+  def calculateSubtotalWithFold(items: List[Item]): Int =
     items
```

Comment:
Example fold: 

```
val res0 = List(1, 2, 3, 4)
res0.foldLeft(0){case (acc, l) => acc + l}
``

Reply:
Cool. I used that as an example in CheckoutOutdated.scala (lines 125-129). I just kept `.sum` because it's essentially a fold, and IntelliJ keeps annoying me! 😆

Reply:
```
def calculateSubtotalWithFold(items: List[Item]): Int = items.map(_.price).sum
```

Can be simplified to:

```
def calculateSubtotalWithFold(items: List[Item]): Int = items.foldLeft(0){case (acc, data) => acc + data.price}
```

This way you dont have to write a map. Good to know for the future :)

Reply:
Ah, great! Mind if I use that...? 😄